### PR TITLE
Upstream ssl

### PR DESCRIPTION
**Pull request by yuval-k**
Allow SSL settings for upstreams.

### CODE REVIEW

**Comment by yuval-k** [comment]
note: this is not yet wired in to gloo
I plan to wire it in and add e2e in a followup pr.

File: projects/gateway/api/v1/virtual_service.proto [hunk 1]
```diff
@@ -9,6 +9,7 @@ import "github.com/solo-io/solo-kit/api/v1/metadata.proto";
 import "github.com/solo-io/solo-kit/api/v1/status.proto";
 import "github.com/solo-io/solo-kit/api/v1/solo-kit.proto";
 
+import "github.com/solo-io/gloo/projects/gloo/api/v1/ssl.proto";
```
Comment (ilackarms) [inline projects/gateway/api/v1/virtual_service.proto]: is this import used here?

File: projects/gloo/pkg/plugins/upstream/plugin.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package upstream
```
Comment (ilackarms) [inline projects/gloo/pkg/plugins/upstream/plugin.go]: suggestion: rename `upstream` to something more specific, eg `upstreamssl` 
referring to this plugin as "the upstream plugin" will get confusing pretty fast

File: projects/gloo/pkg/utils/ssl.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	}
+}
+
+func (s *SslConfigTranslator) ResolveUpstreamSslConfig(uc *v1.UpstreamSslConfig) (*envoyauth.UpstreamTlsContext, error) {
+	if common, err := s.ResolveCommonSslConfig(uc); err == nil {
+		return &envoyauth.UpstreamTlsContext{
+			CommonTlsContext: common,
+			Sni:              uc.Sni,
+		}, nil
+	} else {
```
Comment (ilackarms) [inline projects/gloo/pkg/utils/ssl.go]: nit: no need for `else` here

File: projects/gloo/pkg/utils/ssl.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+		if common.ValidationContextType != nil {
+			requireClientCert = &gogo_types.BoolValue{Value: true}
+		}
+		return &envoyauth.DownstreamTlsContext{
+			CommonTlsContext:         common,
+			RequireClientCertificate: requireClientCert,
+		}, nil
+	} else {
+		return nil, err
+	}
```
Comment (ilackarms) [inline projects/gloo/pkg/utils/ssl.go]: nit: change to 
```
if err != nil { return ...}
return &envoyauth.DownstreamTlsContext{}...
``

File: projects/gateway/api/v1/virtual_service.proto (hunk 1 above)
Comment (yuval-k) [reply projects/gateway/api/v1/virtual_service.proto]: yes; gloo.solo.io.SslConfig ssl_config

File: projects/gloo/pkg/utils/ssl.go (hunk 3 above)
Comment (yuval-k) [reply projects/gloo/pkg/utils/ssl.go]: there is as i use the err variable

**Pull request merged**